use-cases/sports-league-scheduling/src/main/java/org/acme/sportsleagueschedule/solver/SportsLeagueSchedulingConstraintProvider.java

```diff
@@ -0,0 +1,130 @@
+import ai.timefold.solver.core.api.score.stream.ConstraintProvider;
+import ai.timefold.solver.core.api.score.stream.common.SequenceChain;
+
+import org.acme.sportsleagueschedule.domain.Match;
+import org.acme.sportsleagueschedule.domain.Round;
+import org.acme.sportsleagueschedule.domain.Team;
+
+public class SportsLeagueSchedulingConstraintProvider implements ConstraintProvider {
+
+    @Override
+    public Constraint[] defineConstraints(ConstraintFactory constraintFactory) {
+        return new Constraint[] {
+                matchesOnSameDay(constraintFactory),
+                fourConsecutiveHomeMatches(constraintFactory),
+                fourConsecutiveAwayMatches(constraintFactory),
+                repeatMatchOnTheNextDay(constraintFactory),
+                startToAwayHop(constraintFactory),
+                homeToAwayHop(constraintFactory),
+                awayToAwayHop(constraintFactory),
+                awayToHomeHop(constraintFactory),
+                awayToEndHop(constraintFactory),
+                classicMatches(constraintFactory)
+        };
+    }
+
+    protected Constraint matchesOnSameDay(ConstraintFactory constraintFactory) {
+        return constraintFactory
+                .forEachUniquePair(Match.class,
+                        equal(Match::getRoundIndex),
+                        filtering((match1, match2) -> match1.getHomeTeam().equals(match2.getHomeTeam())
+                                || match1.getHomeTeam().equals(match2.getAwayTeam())
+                                || match1.getAwayTeam().equals(match2.getAwayTeam())))
+                .penalize(HardSoftScore.ONE_HARD)
+                .asConstraint("matches on the same day");
+    }
+
+    protected Constraint fourConsecutiveHomeMatches(ConstraintFactory constraintFactory) {
+        return constraintFactory.forEach(Match.class)
+                .join(Team.class, equal(Match::getHomeTeam, Function.identity()))
+                .groupBy((match, team) -> team,
+                        ConstraintCollectors.toConsecutiveSequences((match, team) -> match.getRound(), Round::getIndex))
+                .flattenLast(SequenceChain::getConsecutiveSequences)
+                .filter((team, matches) -> matches.getItems().size() >= 4)
+                .penalize(HardSoftScore.ONE_HARD)
+                .asConstraint("4 consecutive home matches");
+    }
+
+    protected Constraint fourConsecutiveAwayMatches(ConstraintFactory constraintFactory) {
+        return constraintFactory.forEach(Match.class)
+                .join(Team.class, equal(Match::getAwayTeam, Function.identity()))
+                .groupBy((match, team) -> team,
+                        ConstraintCollectors.toConsecutiveSequences((match, team) -> match.getRound(), Round::getIndex))
+                .flattenLast(SequenceChain::getConsecutiveSequences)
+                .filter((team, matches) -> matches.getItems().size() >= 4)
+                .penalize(HardSoftScore.ONE_HARD)
```

Comment:
This is now a score trap. If the result is 5 or 50, it doesn't matter, the penalty will be 1. 

Make the solver's job easier, and penalize by the difference between the minimum and the actual value.